### ISSUE TITLE
Custom sample hooks and descendant spans?

### ISSUE THREAD

**Issue by grncdr** [honeycombio/beeline-ruby#83]
I've implemented a custom sampler that looks like this (in a rails app):

```rb
config.sample_hook do |fields|
  if fields['name'] == 'http_request' && fields['request.path'] == '/some-ignored-path'
    [false, 0]
  else
    [true, 1]
  end
end
```

This does indeed drop the `http_request` spans, but it seems as though all the descendant spans of the request (`process_action.action_controller`, `sql.active_record`, etc. etc.) are still being sent to Honeycomb.

Is this the intended behaviour? I had hoped/assumed that the client would take care of not sending the descendant spans of a dropped span.

**Comment by grncdr** [comment]
I'm not very familiar with the code here, but it seems like the call to `send_children`  here should be moved inside the `if sample ... end` block.

https://github.com/honeycombio/beeline-ruby/blob/344928cbfb44269aed23ab9be099cda79a7bab95/lib/honeycomb/span.rb#L94-L117

**Comment by grncdr** [comment]
Hm, seems that won't solve all issues as child spans created by `start_span() do ... end` are immediately sent:

https://github.com/honeycombio/beeline-ruby/blob/344928cbfb44269aed23ab9be099cda79a7bab95/lib/honeycomb/client.rb#L62-L72

Given this design, I don't think it's possible for a sample hook alone to ensure descendant spans are also dropped, because child spans are sent before the sample hook ever sees the parent span.

**Comment by ajvondrak** [comment]
> Given this design, I don't think it's possible for a sample hook alone to ensure descendant spans are also dropped, because child spans are sent before the sample hook ever sees the parent span.

Not just the design of the sampling hook, but of Honeycomb's event-based system in general. Traces are collections of events. The beeline is going to send spans (which are events) in call stack order, because otherwise it would have to buffer the entire trace before making the sampling decision. It's not impossible, just harder; some preliminary work is being done with [honeycombio/samproxy](https://github.com/honeycombio/samproxy). Buffering would facilitate *tail-based* sampling, where you could use "later" information (like the HTTP status code at the end of your request/response cycle) to decide whether to send the whole trace. This in contrast to the beeline-supported *head-based* sampling, where you can only really make decisions at the start of the root span.

Further reading: 
* https://docs.honeycomb.io/working-with-your-data/tracing/sampling/
* https://www.honeycomb.io/blog/getting-at-the-good-stuff-how-to-sample-traces-in-honeycomb/

For your particular use case, though, you **can** use head-based sampling to discard traces for some specific requests, since you'll know the path ahead of time. That is, you don't need to wait for the whole trace to finish. The problem you're encountering is that `request.path` is only added to the root span - basically using `Honeycomb.add_field`. You could instead add a field to every event that has *not* been sent yet via `Honeycomb.add_field_to_trace`. The trick is adding the trace-level field early enough - before any children get sent.

You might do this by inserting a [middleware](https://guides.rubyonrails.org/rails_on_rack.html). I haven't tested this, but something along the lines of:

```ruby
class RequestSampling
  def initialize(app)
    @app = app
  end

  def call(env)
    Honeycomb.add_field_to_trace('ignore', true) if ignore?(env)
    @app.call(env)
  end

  private

  def ignore?(env)
    env['PATH_INFO'] == '/some-ignored-path'
  end
end
```

Right after Honeycomb starts the root span, your middleware can add the trace-level field.

```ruby
config.middleware.insert_after Honeycomb::Rails::Middleware, RequestSampling
```

Then, your sample hook can look for your trace-level field.

```ruby
config.sample_hook do |fields|
  if fields['app.ignore']
    [false, 0]
  else
    [true, 1]
  end
end
```

Another (possibly simpler) place could be to just call `Honeycomb.add_field_to_trace` in your controller code. But that only works for known paths (which would hit the controller), and I'm not sure how deeply the spans are nested by the point the code reaches the controller anyway. 🤔

**Comment by grncdr** [comment]
Thanks for taking the time to reply Alex. I'm going to respectfully disagree with a couple of your assertions.

> The beeline is going to send spans (which are events) in call stack order, because otherwise it would have to buffer the entire trace before making the sampling decision.

I feel like this is the wrong fit for request/response based applications, but it's a decision with trade-offs. The _potential_ benefit I can see is gettting tracing data for requests that are terminated early, but in that case all you will find in Honeycomb is broken traces.

As shown in this issue, eagerly sending spans also significantly complicates the implementation of accurate sampling. And I haven't even got into trying to hit target rates or other dynamic sampling measures.

In fact it makes it so complicated that Honeycomb is creating an external proxy to handle it. To understand why this doesn't appeal, I'll refer to this section of the [samproxy README](https://github.com/honeycombio/samproxy/blob/ba8a5b766765e6f4f0a14dfd8da930a5faac0c4b/README.md#minimum-configuration):

> ## Minimum configuration
> 
> The Samproxy cluster should have at least 2 servers with 2GB RAM and access to 2 cores each.
> 
> Additional RAM and CPU can be used by increasing configuration values to have a larger CacheCapacity. The cluster should be monitored for panics caused by running out of memory and scaled up (with either more servers or more RAM per server) when they occur.

That's a pretty big ask for alpha software who's primary value (to me) is reducing usage of a commercial service to acceptable levels. Especially when the alternative tradeoff is "buffer spans until the end of the request".

Is there something I'm missing about why buffering spans is a bad idea in practice? I can see the potential for memory bloat, but as far as I can tell shifting that bloat to samproxy isn't much of an improvement.

**Comment by grncdr** [comment]
Sorry, I didn't really address the middleware suggestion you made. As a test I (had already) implemented the head-based sampling and it does cover the basic use case of dropping entire traces. I would still encourage the team to consider allowing tail-based sampling (by sending child spans _only_ from their parent). Or if it's not an acceptable default, exposing APIs to control the sending of spans.

**Comment by ajvondrak** [comment]
> Thanks for taking the time to reply Alex. I'm going to respectfully disagree with a couple of your assertions.

Don't see any disagreements here. I didn't say one approach was better than the other. I just said that buffering is harder. Look no further than the configuration of samproxy, which (as you said) is complicated. You have to make considerations like:
* how much memory are you comfortable letting your unsent traces take up?
* what happens on out-of-memory issues?
* what if a trace terminates in the middle? do we flush the buffer? drop the pending events?
* on what basis do you flush the buffer? timeout? is so, how long?
* what about distributed traces? how do two separate servers know when the same trace ends so they can apply sampling logic? and how do you propagate that decision (or the trace data) between the two servers so that the entire trace can be sent?

Compare to eager event sending, which fires & forgets. :shrug:

> As shown in this issue, eagerly sending spans also significantly complicates the implementation of accurate sampling.

Absolutely. Hence the development of samproxy. Note that I'm not even suggesting you *use* samproxy. I know I wouldn't want that headache.

More feasible would be the [Refinery](https://docs.honeycomb.io/working-with-your-data/tracing/refinery/) product, which takes care of sampling on the (Honeycomb) server side rather than the (beeline) client side. But that's still currently in beta. I'm sure having it server-side probably makes it easier for cross-beeline compatibility, so I understand that architecture from the Honeycomb folks.

> Is there something I'm missing about why buffering spans is a bad idea in practice?

Never said it was a bad idea. Just harder (by which I mean more considerations to make). It's certainly more useful for sampling. AFAICT, having the buffering take place client-side makes distributed trace sampling more of a headache, though.

> ~I can see the potential for memory bloat, but as far as I can tell shifting that bloat to samproxy isn't much of an improvement.~
>
> Edit: The comment about memory bloat is irrelevant, as the beeline is already retaining references to all child spans in the `Span#children` array.

Hey, that's an interesting point! I personally find the structure of the beeline code hard to follow, and this was one of the reasons. As-is, that array is only kept around in case the children manage not to send themselves like they're supposed to. [This discussion](https://honeycombpollinators.slack.com/archives/CJR134U2F/p1579116563027100) in the Pollinators Slack is relevant. I'm not sure when such a case would present itself, though. 🤔  Maybe a non-`StandardError` exception, so `Honeycomb::Client#start_span` won't rescue that?

But you're right, the memory exists. Apparently the answer to several of the above considerations is "let it grow". 😅

**Comment by grncdr** [comment]
Sorry if I came off as combative, it wasn't clear to me whether or not you were a Honeycomb employee and making recommendations in that capacity or (as it seems) just another interested user.

Your points about distributed traces are well-taken, but I don't think they have an impact on this particular question about the internal design of the Beeline. Regardless of when the beeline sends spans, propagating a sampling decision across process boundaries requires an outside perspective (e.g. samproxy or Refinery).

The issue I'm raising is only about propagating sampling decisions within a single `Honeycomb::Client` instance. Limited to that scope, the questions you put forward all have a relatively simple answers.

> how much memory are you comfortable letting your unsent traces take up?

Apparently as much as they currently do 😉 

> what happens on out-of-memory issues?

Bad things 😄 . I can definitely see a use-case for a `Honeycomb.silent do ... end` type of API so that an integration doesn't generate a huge amount of tracing data in a tight loop somewhere.


> on what basis do you flush the buffer? timeout? is so, how long?

When a `root`/ `subroot` span ends and the sample hook returns an affirmative response, call `send_children`, then send the parent event. This will already trigger recursive `send_children` calls which walk the trace tree and allow the sampling decisions to be made at each level, so that uninteresting branches of a trace can be skipped entirely.

> what if a trace terminates in the middle? do we flush the buffer? drop the pending events?

If you mean when an exception is caught, then behave exactly the same as regular execution. Run the sample hook, send children, send the parent. In the case where there is no exception, (e.g. process is killed without cleanup) this design will lose more data than one that eagerly sends spans. However, IME these are relatively rare occurences, and application code could still explicitly call `.send` on a child span in code paths where this is deemed likely.

> what about distributed traces? how do two separate servers know when the same trace ends 
> so they can apply sampling logic? and how do you propagate that decision (or the trace data) > between the two servers so that the entire trace can be sent?

As mentioned above, distributed traces are an orthogonal issue to what I'm talking about here.

Disclaimer: I'm not a yet member of the Honeycomb Slack, so I couldn't read the linked discussion. Apologies if I'm rehashing arguments that have already been made.

**Comment by irvingpop** [comment]
Hey @grncdr I understand your struggle here - sampling becomes so much more complex when tracing is involved. 

The #1 rule of Beelines is that they absolutely cannot impact the performance of an application, and that's why we may never implement tail-based sampling in the beelines - it adds significant computation and memory expense that are simply not acceptable to our users. 

As I wrote in the blog post Alex mentioned, all we can do is move that cost around - ideally outside of the application process.  One option we evaluated was the Refinery closed-beta, but we aren't accepting new users to that program right now.  Current options include Honeycomb samproxy (still in Alpha) and the rapidly evolving [OpenTelemetry collector](https://www.honeycomb.io/blog/opentelemetry-new-honeycomb-exporters/)

We know that our users sampling options aren't ideal right now - head-based sampling has many limitations, broken traces really suck. Our ideal sampling solution isn't written yet, but we are actively thinking about it and will be improving it.

**Comment by ajvondrak** [comment]
@grncdr: 

> Sorry if I came off as combative, it wasn't clear to me whether or not you were a Honeycomb employee and making recommendations in that capacity or (as it seems) just another interested user.

Ah, fair. 😅 Just a concerned citizen like you, I'm afraid. I appreciate the back & forth, though, so no worries. 🙃

As far as design goes, what you present is elegant in its own way. Just like eager-sending is elegant in the opposite way. 😛 In my mind, the circular design of `Honeycomb::Context#spans` and `Honeycomb::Span#children` is really not picking which lane it wants to stay in: either send events eagerly or don't. I wonder if I'm missing an obvious case where `meta.sent_by_parent` could actually come up in practice.

For the simple non-distributed trace use case, what you outline is pretty straightforward to implement in a way that's just as memory (in?)efficient as the current beeline is today. 🤔 It'd be feasible to present it as a configuration option, I suppose, but the structure of the code is different enough that I think it'd amount to dependency-injecting a choice between two different libraries anyway. But I have a lot of ideas about that, lol. I've pondered alternative beeline interfaces quite a bit. Not that users should need to implement entire libraries themselves, but the flexibility exists.

Ultimately, I do understand where Honeycomb & @irvingpop are coming from. Centralizing server-side sampling services (say that ten times fast) like samproxy, Refinery, or the otel collector elegantly solves some of their more general issues: removing the client-side costs, handling distributed traces, making tracing more user-friendly (wire up your code _just so_ vs toggle some switches in the Honeycomb UI), not having to implement tail-based sampling across all their beelines (which may not be as cavalier about memory usage as Ruby - I'm just thinking of my Java days of having to tune the JVM's heap size 😓), etc.

**Comment by grncdr** [comment]
> that's why we may never implement tail-based sampling in the beelines - it adds significant computation and memory expense that are simply not acceptable to our users.

Apologies if I've misunderstood, but I think the changes outlined in my previous comment could only *reduce* the performance impact of the Beeline. Furthermore, when referring to the Rack middleware, that performance impact would be deferred until *after* a response has been sent, which would be pretty nice.

| |current "eager send"|proposed "lazy send"|
|-|-------|---------|
|memory|all spans are retained until the root span is sent|all spans are retained until the root span is sent|
|cpu|sample hook is called after every span during the request|sample hook is called for root span and any sampled descendants, after response is sent|

**Comment by martin308** [comment]
@grncdr perhaps I can provider some further explanation: 

The fact that spans are not removed the children array is a bug in the ruby beeline. It is not the intention that these spans are retained for the duration of the parent span. You can see the piece of code that is missing from the ruby beeline in the [go beeline](https://github.com/honeycombio/beeline-go/blob/master/trace/trace.go#L321). I will create another issue to track this bug.

Additionally we are striving for uniformity across all of the beelines, if we wanted to implement this method of sampling in the ruby beeline we would have to add/support this across all of the other languages that we support. As @irvingpop mentioned we know that our sampling options aren't ideal right now and it is something that we are working on and thinking about. 

Thanks all for the in depth discussion!
